src/wallet/spend.cpp

```diff
@@ -620,15 +630,59 @@ static void DiscourageFeeSniping(CMutableTransaction& tx, FastRandomContext& rng
     // now we ensure code won't be written that makes assumptions about
     // nLockTime that preclude a fix later.
+    //
+    // nSequence-based
+    // ---------------
+    //
+    // If all inputs in the tx spend taproot coins, use nSequence to discourage
+    // fee-sniping. This creates a "privacy cloak" for txs that use taproot
+    // key-path spends (BIP326).
+    //
+    // Compared to nLockTime-based anti-fee-sniping, this may be minimally less
+    // effective, as nSequence only allows relative locks. If all inputs of the
+    // tx can be moved backward in the chain, this tx can be as well. Though,
+    // this should be fine given that 10% of txs have their locktime randomly
+    // reduced to a past block. Also, sequence-based locking will only be used
+    // if every sequence is at least 1.
     if (IsCurrentForAntiFeeSniping(chain, block_hash)) {
-        tx.nLockTime = block_height;
-
-        // Secondly occasionally randomly pick a nLockTime even further back, so
-        // that transactions that are delayed after signing for whatever reason,
-        // e.g. high-latency mix networks and some CoinJoin implementations, have
-        // better privacy.
-        if (rng_fast.randrange(10) == 0) {
-            tx.nLockTime = std::max(0, int(tx.nLockTime) - int(rng_fast.randrange(100)));
+        auto locktime_based{[&] {
+            if (!SignalsOptInRBF(CTransaction{tx})) {
+                // Use locktime if the tx inputs do not opt-in to RBF
+                return true;
+            }
+            for (const auto& in : inputs) {
+                if (in.depth < 1 || uint64_t(in.depth) > MAX_BLOCK_SEQ) {
+                    // Use locktime if any depth is out-of-range
+                    return true;
+                }
+                std::vector<std::vector<uint8_t>> dummy;
+                const TxoutType in_type{Solver(in.txout.scriptPubKey, dummy)};
+                // Use locktime if any input isn't taproot
+                if (in_type != TxoutType::WITNESS_V1_TAPROOT) {
+                    return true;
+                }
+            }
+            return false;
+        }};
+        if (locktime_based() || rng_fast.randrange(2) == 0) {
+            tx.nLockTime = block_height;
+
+            // Secondly occasionally randomly pick a nLockTime even further back, so
+            // that transactions that are delayed after signing for whatever reason,
+            // e.g. high-latency mix networks and some CoinJoin implementations, have
+            // better privacy.
+            if (rng_fast.randrange(10) == 0) {
+                tx.nLockTime = std::max(0, int(tx.nLockTime) - int(rng_fast.randrange(100)));
+            }
+        } else { // sequence based
+            tx.nLockTime = 0;
+            const auto i_input{rng_fast.randrange(tx.vin.size())};
+            CTxIn& in{tx.vin.at(i_input)};
+            in.nSequence = inputs.at(i_input).depth;
```

Comment:
There could be some unexpected behavior here when users set their own sequence numbers.

If a user set their own sequence, we would not want to clobber it as that could have negative effects on their payment. So we should avoid changing the sequence on any inputs that use a non-default sequence number.

However, `DiscourageFeeSniping` is called by `CreateTransactionInternal` which does not actually set (or know about) user defined sequences. User defined sequences are handled externally in `FundTransaction` where the it actually retains the `CTxIn`s of the transaction to fund and then adds the inputs that `CreateTransaction` chooses.

So user defined sequences actually end up in the transaction and we lose the anti-fee-sniping that we wanted.

I think the correct way to do this is to make `CreateTransactionInternal` aware of user defined sequences and set them in the transaction it is creating. Then here in `DiscrouageFeeSniping`, we can check just choose randomly among the inputs that have default sequences.

However, what should we do if the user sets a default sequence explicitly? Should we respect that and avoid randomly selecting those inputs? Additionally, even if a user did set a default sequence explicitly, `fundrawtransaction` would not know because it takes a raw transaction which has sequences that `createrawtransaction` may have set using defaults.

Reply:
I think we should never fiddle with user chosen sequence numbers, even if they happen to be "default". I also think this is the way that createrawtransaction/createpsbt works (it wouldn't set locktime nonzero unless the user asked for it and it wouldn't set sequence in violation of the `replaceable` boolean option unless the user asked for it).

If you are asking to extend fee sniping to createrawtransaction/createpsbt, I think this should be done in a separate pull request.

The sequence numbers in CreateTransactionInternal solely depend on `const bool use_rbf`, so I think the changes here are both safe and sane.

Reply:
Oh, good catch. Though, it appears this bug already exists in current master.

You can test this by calling `createrawtransaction` with `replacable=True` and providing some inputs, then have `fundrawtransaction` pick the remaining inputs. Now anyone can tell apart the `createrawtransaction`-inputs from the `fundrawtransaction`-inputs by looking at the sequence number.